OpenSim/Simulation/MarkersReference.cpp

```diff
@@ -22,139 +22,157 @@
-    // pre-allocate arrays to the number of markers in the file with default weightings
-    _markerNames.assign(nm, "");
-    _weights.assign(nm, get_default_weight());
+    _markerNames.assign(markerNames.size(), "");
+    _weights.assign(markerNames.size(), get_default_weight());
 
-    for (int i = 0; i < nm; i++) {
-        const std::string &name = tempNames[i];
-        _markerNames[i] = name;
+    for(unsigned i = 0; i < markerNames.size(); ++i) {
+        _markerNames[i] = markerNames[i];
     }
 
     // Names must be assigned before weights can be updated
     updateInternalWeights();
 }
 
-SimTK::Vec2 MarkersReference::getValidTimeRange() const
-{
-    return Vec2(_markerData->getStartFrameTime(), _markerData->getLastFrameTime());
+SimTK::Vec2 MarkersReference::getValidTimeRange() const {
+    OPENSIM_THROW_IF(_markerTable.getNumRows() == 0,
+                     Exception,
+                     "Marker-table is empty.");
+
+    return {_markerTable.getIndependentColumn().front(),
+            _markerTable.getIndependentColumn().back()};
 }
 
-// utility to define object properties including their tags, comments and 
-// default values.
-void MarkersReference::constructProperties()
-{
+void MarkersReference::constructProperties() {
     constructProperty_marker_file("");
     Set<MarkerWeight> markerWeights;
     constructProperty_marker_weights(markerWeights);
     constructProperty_default_weight(1.0);
 }
 
-/** get the names of the markers serving as references */
-const SimTK::Array_<std::string>& MarkersReference::getNames() const
-{
+const SimTK::Array_<std::string>& MarkersReference::getNames() const {
     return _markerNames;
 }
 
-/** get the values of the MarkersReference */
-void  MarkersReference::getValues(const SimTK::State &s, SimTK::Array_<Vec3> &values) const
-{
-    double time =  s.getTime();
-
-    int before=0, after=0;
-    // get index for time 
-    _markerData->findFrameRange(time, time, before, after);
-    if(before > after || after < 0)
-        throw Exception("MarkersReference: No index corresponding to time of frame.");
-    else if(after-before > 0){
-        before = abs(_markerData->getFrame(before).getFrameTime()-time) < abs(_markerData->getFrame(after).getFrameTime()-time) ? before : after;
-    }
-
-    values = _markerData->getFrame(before).getMarkers();
+void  MarkersReference::getValues(const SimTK::State& s,
+                                  SimTK::Array_<Vec3>& values) const {
+    double time = s.getTime();
+    const auto rowView = _markerTable.getNearestRow(time);
+    values.clear();
+    for(unsigned i = 0; i < rowView.ncol(); ++i)
+        values.push_back(rowView[i]);
 }
 
-/** get the speed value of the MarkersReference */
-void MarkersReference::getSpeedValues(const SimTK::State &s, SimTK::Array_<Vec3> &speedValues) const
-{
+void
+MarkersReference::getSpeedValues(const SimTK::State &s,
+                                 SimTK::Array_<Vec3> &speedValues) const {
```

Comment:
Is there a reason to keep the unimplemented method here? A user looking at documentation will get the false assumption that it's implemented only to be surprised at runtime. I'd prefer to remove it unless there's work in progress to implement it (short term) or this class implements an interface that requires the method and there are other instances where  implementation actually exist. I agree it will be great when implemented but don't see benefit to users.